### ISSUE TITLE
Implement simple chat functionality between two separate users.

### ISSUE THREAD

**Issue by Jldevictoria** [Victory-Game-Studios/Oldentide#27]
Users can now log in thanks to Michael.  Now we want to let them send messages among other players on the server.

These messages are sent through the PACKET_PLAYERCOMMAND packet (in the command portion) and each client should have a threaded listener in charge of waiting for packets from the server to arrive.

The server will need to handle parsing packets from the users and relaying their information to all of the involved parties. 

These are the types of channels we will be using:

/s - Say command. 
This type of command is used to speak to anyone within 100 in game meters from the user.  We will need to calculate the distance of all people in the zone where the user is and relay the message to those players.  This is a required channel.

/y - Yell command.
This command is used to speak to anyone within 300 in game meters from the user.  Again we will need to calculate the distance from all people in the zone where the user is and relay the message to those players.  This is meant to have a more urgent feel.  This is a required channel.

/ooc - Out of character command.
This channel is zone wide, and is meant to be used for casual chat between players, esp on Role-play servers.  This server administrator should eventually be able to disable this channel.  If this channel is disabled, the server should send the client a server message indicating that it is disabled.

/h - Help command.
This channel is a server wide channel meant for GMs and players to help new players.  Generally this would be a moderated channel.  The server administrator should be able to disabled and enable this channel at will.  If this channel is disabled, the server should send the client a server message indicating that it is disabled.

/w - Whisper command.
The whisper command should relay messages server wide to a designated player.  For example, if I do a "/w Fred Hey there man what's up?"  The server should recognize a whisper command, find Fred from among the users and send him my message.  Fred should see "Hey there man what's up?" on his screen and nobody else.  If Fred is not playing then the sender should receive a server message indicating that no such player is logged on.  This is a required channel.

/g - Guild command.
This channel should relay messages server wide to any members of a given guild.  This is a required channel.

/p - Party command.
This channels should relay messages server wide to any player within a given party.  This is a required channel.

This may be adjusted if needed!

**Comment by hintron** [comment]
Currently, each client has a separate thread that regularly polls the server and syncs any chats. I made it this way to temporarily get a simple chat service up and running. It's a short-term solution.

Eventually, we want the server to be able to broadcast messages to all clients without the clients needing to poll the server every second. This will require the server to at least: 

- know who is all connected (so it knows who it can broadcast to and what address they are at, as well as routing specific messages like whisper)

- have a timeout or connection test function, to know when clients are no longer playing (and maybe a client can send a disconnect, so this will only be needed if a client crashes or loses internet connection)

Changing things in this way will make it so that we won't need to store all messages on the server - we just simply relay them and then discard afterwards.

**Comment by Jldevictoria** [comment]
One thing that I have though about doing is to start using the PACKET_ACK to acknowledge packet transmission and place a timeout based on the ACK response.  For example:

Player sends command: "/w Tom Come to the city quick!" with packetId 1350.

Server sees packet and sends back a PACKET_ACK with the packetId 1350 to acknowledge recpetion of the /w command.

Player receives Ack packet and registers his chat message as sent.

Server sends off PACKET_SERVERCOMMAND to Tom's client with packetId 2034 relaying the first player's message.  Server creates a listener thread waiting for Tom's Ack packet to return.

tom's client has disconnected!  So no Ack packet is returned for packetId 2034 on Tom's session!

Server listener thread times out after 6 seconds or so and marks Tom's client as disconected.

Player tries to send another message: "/w Tom are you there? we need you!"

Server receives packet, parses it and sees that Tom's client is no longer connected, sends a PACKET_SERVERCOMMAND to Player indicating that Tom has disconnected and his message was not transmitted.

Something like that would not be terribly hard to implement I think.

**Comment by Jldevictoria** [comment]
Let's focus on this as a next step for Oldentide.
I think we have everything we need to implement simple chat. 

Lets flesh out the PLAYERCOMMAND handler on the Server so that it broadcasts every message that it receives to every player that is currently logged in. (We should be saving the connection information for each player in the Gamestate)

Then I will create a simple interface on the Client that offers a command prompt to the user.  I will add functionality for the "/s" command that will simply send the PLAYERCOMMAND packet with the payload "/s <message>"

I will also add a listener to the Client that will pick up incoming packets and handle the PLAYERCOMMAND "/s" packets by printing their contents to the screen.

Eventually we can add Player names as well.

**Comment by hintron** [comment]
I agree. The next thing I plan on looking at is connecting the Unity Client to the server via message pack, which shouldn't be too hard because I already got it working once in NetworkManager.cs - I just need to add in our new custom header stuff. But as it is right now, the Unity Client won't be able to talk to the server.

While I figure that out, though, you could be building the infrastructure to get two TestClients speaking to each other. These two things are independent of each other.

**Comment by hintron** [comment]
I've made a lot of progress - the Unity Client now talks to the server properly, in a non-blocking way. Now I'm just kludging together a simple server broadcast function that will send a string to all connected clients. Then, I'll create a non-blocking listener thread in Unity that will just wait for ServerSendCommand packets sent from the broadcast function.

**Comment by hintron** [comment]
Ok, the basic messaging functionality now works with both TestClient and Unity client. In the Unity client, press "m" to connect to the server, and press "i" to broadcast a hard-coded message. In the server admin shell or in a TestClient, "/h My broadcast" will broadcast a message to all connections, and "/s 3 Hi client with sessionId 3" will send a personal message. You can also send personal messages to the admin shell: "/s admin Hi admin"

The next step is to obviously get user input from the Unity client. How should we do that? Create a UI console/messages popup? I think @Jldevictoria should have a better idea of what to do here.

**Comment by hintron** [comment]
So, I'm running into a kind huge problem... [Unity is not built to work with multiple threads](http://stackoverflow.com/questions/41330771/use-unity-api-from-another-thread-or-call-a-function-in-the-main-thread). Yeah... What is happening to me right now is that I am able to get data from the server in unity using my threads, but I can't forward that data to the main unity thread to update... When I do, I get a "can only be called from the main thread" error. See the attached picture. I should have known this approach wouldn't work, because Unity isn't thread safe.

The good news, though, is that it looks like the 'asyc' socket calls can work, and I played around with them and got them working at one point. That just might have to be the way we do it. I'll need to think it through to see if I can implement a packet queue architecture using the async calls.

![image](https://cloud.githubusercontent.com/assets/9166578/26022987/7b1b2288-3767-11e7-804c-bc24667461e3.png)

**Comment by hintron** [comment]
So, the async method also emits a thread warning... which makes sense, because I've read that the async callbacks are executed in a separate thread. So back to the drawing board.

[There is another method that looks promising though](http://answers.unity3d.com/questions/200176/multithreading-and-messaging.html). I think the basic idea is that you send code snippets between threads, so that the main thread executes the code you wanted to execute in the worker thread. That way, you don't violate unity's outside thread API access policy. I'll look into it.

**Comment by hintron** [comment]
Update: I got data from the server being passed to the Unity client very well now! It also doesn't block if it sends out packets that aren't responded to by the server. See pictures.

So I ended up doing a kind of state machine architecture with packets. There is still a small second listener thread (via an async callback to the beginReceive socket call), so it doesn't block Unity. However, all it does is wait for a packet, obtain a lock on the packet queue and enqueue a packet. Then, the Update() function in the main unity thread will obtain a lock on the packet queue, check if count >0, and dequeue a packet to work on if there are any. The locking will never block Unity longer than it takes to copy the packet, so effectively not at all.

![unity_messages_in_client_2](https://cloud.githubusercontent.com/assets/9166578/26028996/9c7dd538-37df-11e7-902d-414381e71932.PNG)
![unity_messages_in_client_3](https://cloud.githubusercontent.com/assets/9166578/26028995/9c7ae670-37df-11e7-9eae-0b39b8380f4a.PNG)

**Comment by Jldevictoria** [comment]
@hintron proved that this can be done.  I will be updating the client to handle these packets once I get an in-game terminal working.
